### ISSUE TITLE
Difficulties setting rungs / stopping

### ISSUE THREAD

**Issue by amueller** [awslabs/syne-tune#777]
Guess who it is? Hopefully my questions & comments are helpful for you to figure out how to improve usability. I'm super stoked about this tool, so I'm trying to give you some feedback. If this is not useful, let me know and I'll try to figure out things by myself more.

I started running my "real" experiments, using MOBSTER and ``type="promotion"``. Currently my main issue is #773 since I'm not sure how to reuse results, but apart from that, I'm not getting the results I'm looking for, and I think it's because I didn't set the stopping criterion right (or more generally I set up the rungs badly).

I'm running on 4 GPUs, the config space has ``'epochs': 4000``, the tuner looks like this:
```
tuner = Tuner(
    trial_backend=LocalBackend(entry_point='../fit_model.py'),
    scheduler=MOBSTER(
        config_space,
        metric='loss',
        resource_attr='epoch',
        max_resource_attr="epochs",
        search_options={'debug_log': False},
        mode='min',
        type="promotion",
        grace_period=5,
    ),
    max_failures=1000,
    results_update_interval=60,
    print_update_interval=120,
    stop_criterion=StoppingCriterion(max_num_trials_started=2000),
    n_workers=4,  # how many trials are evaluated in parallel
    tuner_name=tuner_name,
    trial_backend_path=f"/datadrive/synetune/{tuner_name}/"
)
```
and the results look like this:
<img width="526" alt="image" src="https://github.com/awslabs/syne-tune/assets/449558/39009b13-faff-4a8f-85dd-74824dc9a78c">

So it looks to me like we never get out of the first rung, but I'm also a bit confused by the promotion.
All the jobs were started with ``epochs=5`` and some were paused and later resumed. I had assumed that everything would be started with ``epochs=4000`` and paused only after 5 epochs.
I assume the rungs get assigned based on the ``max_num_trials_started=2000`` and so promotion will only occur after ``2000/3`` trials have been started?
Having a guide on how to pick rungs and/or stopping criteria would be great. In an ideal world, I'd like to have an any-time algorithm that keeps improving but doesn't take forever until I get reasonable results. I'm not sure if any of the algorithms you implement (or any of the configuration options) achieve that.

I'm also a bit confused about how promotion happens (and I might need to just read the docs and examples more carefully). Since everything got started with 5 epochs, the runs actually ended after 5 epochs, so in my code, there's no way to restart / continue them, so once we get to the next rung, we have to start again from scratch. That's not how I understood the promotion to happen from my vague idea from @mseeger's talk.

**Comment by amueller** [comment]
Reading ``https://syne-tune.readthedocs.io/en/latest/tutorials/basics/basics_promotion.html``, if I remove ``max_resource_attr``, I should get what I want and the runs will all be started with ``epochs=4000`` and the scheduler will pause them when appropriate?

**Comment by amueller** [comment]
Hm, setting ``max_number_of_trials`` lower seems to create a system that stops really early without ever allocating ``max_resources`` to any run: ``syne_tune.stopping_criterion:reaching max number of trials started (37), stopping there``

**Comment by amueller** [comment]
Ok I think I found my mistake. I set ``max_resource_attr`` to ``epochs``, but really for the promotion case, I needed to add another parameter "stop after epochs" that basically stops the training loop after that many epochs, and assign that to ``max_resource_attr``, while just statically setting ``epochs`` to the maximum number of epochs I want to assign to each trial. That makes the learning rate scheduling and load/resume logic work as expected, I think, but it's slightly counter-intuitive.
lmk if I'm making any sense.

**Comment by jgolebiowski** [comment]
Hey, I am wondering if there is something else we can do for this issue?
Apologies for the delayed response.
Jacek

**Comment by amueller** [comment]
Maybe having a more clear guide on how to implement the promotion case and what parameters are expected and how the stopping and resume is done would be good?
Actually I think in my last comment, I was still not clear on how it works.
So it seems like the training job is not actually responsible for stopping itself in the promotion case, but instead it's killed.
So I'm not sure what the ``resource_attr`` is doing in the promotion case, if anything?

**Comment by jgolebiowski** [comment]
Ok makes sense, I appreciate the documentation there can be somewhat confusing, I will try to make the doctrings clearer so this stops being an issue.

**Comment by mseeger** [comment]
Hello Andreas, I'll look into this. What you observe is correct. If you run a pause-and-resume scheduler and start a trial which is supposed to run maximally for 2000 epochs, but the first level it is paused is at 5 epochs, then the training task is started to be run until 5 epochs (and not 2000).

The reason is that we want to avoid overhead having to stop the trial. We know it has to pause after 5 epochs, so we just tell it to stop then. This does NOT mean it cannot be resumed.

In fact, this allows us to run things with (hypothetical) backends which cannot stop trials from the outside.

**Comment by mseeger** [comment]
BTW: Thanks for using Syne Tune on your MotherNet work. If there is any paper or blog post coming out of this, please tell us, so we can link it from our docs. Also, maybe we can have a smaller variant of your tuning problem as an example?

**Comment by mseeger** [comment]
OK, I think your questions are somehow answered in the multi-fidelity tutorial, but one may never get to this. I'll add an entry to FAQ.

In a sense, you are right, that for a pause-and-resume scheduler, we do control exactly how long a trial job runs, so it would not have to report epoch numbers.

But in practice, we want people to implement a single training script which works for different tuning algorithms, not just for pause-and-resume. This is why we require to report epoch numbers.

**Comment by mseeger** [comment]
I hope #809 closes this issue. If not, feel free to open it again.

**Comment by amueller** [comment]
@mseeger thank you for the responses and welcome back! The current paper doesn't include synetune-based results yet, but the next iteration probably will. Also, it seems likely now that some work will be opensourced, which means we could create an example from it.
Maybe add in the FAQ that in the promotion case, the jobs are killed and `resource_attr` isn't used?

Is there a reason to do pause-and-resume in a multi-worker setup instead of promotion?

As a used, I probably don't really care about trying out many different algorithms. It takes like a couple of weeks and like $10k or so to run one, so I'm probably going to focus on those that seem most promising, and adding infrastructure to support all possible algorithms is some overhead. I know you have a slightly different scenario since you're trying to benchmark the tuning algorithms, but I certainly won't do that on this problem.

**Comment by mseeger** [comment]
Well, it is not that the jobs are killed, but they are simply run not for longer than needed, and terminate on their own, so the backend does not have to kill them (just to be safe, the backend right now still tries to kill such jobs, but they'll have finished on their own).

Promotion-based hyperband is a pause-and-resume scheduler. Pause-and-resume just means that a trial (linked to a config) can be paused and later resumed, e.g. from a checkpoint. In contrast, early stopping schedulers once they stop a trial, they cannot resume it.

**Comment by mseeger** [comment]
I think what you do, to use HPO as an interactive tool and distribute resources at a higher level, makes a lot of sense, I always thought about it that way.

**Comment by amueller** [comment]
I guess I was confused by there being two mechanisms, because it seemed that I don't have to implement any stopping logic if the process is just killed from the outside. In the other thread you said it'll be hard to implement binned time in the training script, but if I just rely on the job being killed, then I don't need to worry about it in the training script, right?

**Comment by mseeger** [comment]
Well, you need to worry about reporting the metrics at resource level r=1, 2, 3, 4, ... And if you use time, this means you need to report them (roughly) after X, 2 X, 3 X, 4 X, ... seconds, where X is the grace period you choose.

This is not very hard to do, but it is harder than reporting just at the end of each epoch. It needs you to keep track of time, and check (say) after each mini-batch update whether you crossed the next k*X point.

**Comment by mseeger** [comment]
But if this is useful for you, let's do it!

**Comment by amueller** [comment]
Yeah I think that's what I've been doing (roughly).

**Comment by mseeger** [comment]
OK, then you could try immediately how it works. You'd just define a map from `float` to `int`. For regular spacing, this could be as simple as `ceil(time / width)`, and you report that as `resource_attr`.

If you implement checkpointing, make sure to store the (total) time until the CP is stored, and start from this time (as offset) when you resume.

**Comment by amueller** [comment]
Is missing bins fine? Right now in the code you're not allowed to overshoot, right?

**Comment by mseeger** [comment]
There should not be gaps, correct.
